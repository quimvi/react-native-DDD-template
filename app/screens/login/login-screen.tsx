import { useNavigation } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { observer } from 'mobx-react-lite';
import React from 'react';
import { View, Button } from 'react-native';

import { useApplicationLayer } from '~/hooks';
import { translate } from '~/i18n';
import { RootParamList } from '~/navigation/params';

import { LoginScreenProps as Props } from './login-screen.props';
import { LoginScreenStyles as styles } from './login-screen.styles';

type LoginScreenNavigationProp = StackNavigationProp<RootParamList, 'Login'>;

export const LoginScreen: React.FC<Props> = observer(() => {
  const {
    useCases: { Login },
  } = useApplicationLayer();
  const navigation = useNavigation<LoginScreenNavigationProp>();

  function handleLogin() {
    Login().subscribeToRequest({
      onCompleteRequest: () => {
        navigation.navigate('Main');
      },
    });
  }

  return (
    <>
      <View style={styles.view.content}>
        <Button
          title={translate('common.sendButton')}
          onPress={() => handleLogin()}
        />
      </View>
    </>
  );
});
