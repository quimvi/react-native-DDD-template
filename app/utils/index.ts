/**
 * Parse a string to boolean.
 *
 * @param {string} bool String to check.
 */
export function parseBoolean(bool: string) {
  return ['1', 'True', 'true'].includes(bool);
}
