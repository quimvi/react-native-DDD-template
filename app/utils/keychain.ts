import * as ReactNativeKeychain from 'react-native-keychain';

/**
 * Saves some credentials securely.
 *
 * @param username The username
 * @param password The password
 * @param server The server these creds are for.
 */
export async function save(username: string, password: string, server?: string) {
  if (server) {
    await ReactNativeKeychain.setInternetCredentials(server, username, password);
    return true;
  } else {
    return ReactNativeKeychain.setGenericPassword(username, password);
  }
}

/**
 * Loads credentials that were already saved.
 *
 * @param server The server that these creds are for
 */
export async function load(server?: string) {
  if (server) {
    const creds = await ReactNativeKeychain.getInternetCredentials(server);
    return {
      password: creds ? creds.password : null,
      server,
      username: creds ? creds.username : null,
    };
  } else {
    const creds = await ReactNativeKeychain.getGenericPassword();
    if (typeof creds === 'object') {
      return {
        password: creds.password,
        server: null,
        username: creds.username,
      };
    } else {
      return {
        password: null,
        server: null,
        username: null,
      };
    }
  }
}

/**
 * Resets any existing credentials for the given server.
 *
 * @param server The server which has these creds
 */
export async function reset(server?: string) {
  if (server) {
    await ReactNativeKeychain.resetInternetCredentials(server);
    return true;
  } else {
    const result = await ReactNativeKeychain.resetGenericPassword();
    return result;
  }
}
