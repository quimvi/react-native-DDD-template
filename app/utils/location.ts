export function toRadians(value: number): number {
  return (value * Math.PI) / 180;
}

export const EARTH_RADIUS_METERS = 6371000;
