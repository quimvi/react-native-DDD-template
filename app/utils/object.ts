import { mergeAll, flatten, reduce, assoc, keys } from 'ramda';

/**
 * Search in an object the key corresponding to a value
 *
 * @param obj Object where we are going to search the key
 * @param value Value to search
 */
export function getKeyByValue(obj: object, value: any) {
  return Object.keys(obj).find((key) => obj[key] === value);
}

/**
 * Creates a list of string from an object, to represent all the keys
 * with the format ['key1.key2','key1.key3']
 *
 * @param obj Object to generate the list
 * @param keyPrefix Add a prefix to each key
 * @param separator Separator to use between keys
 */
export function extractKeys(obj: object, keyPrefix = '', separator = '.'): string[] {
  const combinedKeys = [];
  const keys = Object.keys(obj);

  keys.forEach((key) => {
    if (typeof obj[key] === 'string') {
      if (key.includes('_plural')) {
        return;
      }
      combinedKeys.push(keyPrefix + key);
    } else {
      combinedKeys.push(...extractKeys(obj[key], keyPrefix + key + separator));
    }
  });

  return combinedKeys;
}

/**
 * Creates a new object formed by the keys which has the attribute passed.
 * Each key value will be the value of that attribute.
 *
 * @param obj Target object which has the desired attribute
 * @param attribute Attribute to return.
 */
export function reduceKeyByAttribute(obj: object, attribute: string) {
  return Object.keys(obj).reduce(function (accum, key) {
    const item = obj[key];

    if (attribute in item) {
      accum[key] = item[attribute];
    }

    return accum;
  }, {});
}

/**
 * Mix two or more objects.
 *
 * @param {object[]} objects The array of objects to mix.
 */
export const enhance = <T extends object>(objects: object[]): T => {
  return mergeAll(flatten(objects)) as T;
};

/**
 * If element is an array, return the element[index]. If not, return element
 * @param element Array or default value for the preset.
 * @param index Item to access in case preset is an array.
 */
export function getElement<T>(element: T | T[], index: number): T {
  if (Array.isArray(element)) {
    return element.length > index ? element[index] : null;
  } else {
    return element;
  }
}

/**
 * Given an array an a map of key:string -> value:string, rename the key properties in each
 * object in the array for the value.
 * @param array Array to rename
 * @param keysMap Map of renamings
 */
export function renameKey<T>(array: T[], keysMap: Record<string, string>): T[] {
  return array.map((obj: T) =>
    reduce((acc, key) => assoc(keysMap[key] || key, obj[key], acc), {} as T, keys(obj)),
  );
}
