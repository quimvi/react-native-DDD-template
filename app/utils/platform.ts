import { Platform } from 'react-native';

/**
 * Returns the Android value if the platform is Android, and the ios value if the platform is iOS
 * @param android Value to return if is Android
 * @param platform Value to return if is iOS
 */
function select<T>(android: T, ios: T): T {
  return Platform.select({
    android,
    ios,
  });
}

export default { select };
