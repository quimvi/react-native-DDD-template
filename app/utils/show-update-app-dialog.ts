import { Alert, BackHandler } from 'react-native';

import { translate } from '~/i18n';

export function showUpdateAppDialog() {
  Alert.alert(
    translate('component.titleMandatoryUpdate'),
    translate('component.subtitleCloseApp'),
    [
      {
        onPress: () => {
          BackHandler.exitApp();
          return false;
        },
        style: 'cancel',
        text: translate('button.closeApp'),
      },
    ],
    { cancelable: false },
  );
}
