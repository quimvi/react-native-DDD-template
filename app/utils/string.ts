/**
 * Capitalize first word of the string
 *
 * @param {string} word String
 */
export function capitalize(word: string) {
  return typeof word === 'string' ? word.replace(/^\w/, (w) => w.toUpperCase()) : '';
}

/**
 * Capitalize all words of a string
 *
 * @param {string} words String containing the words
 */
export function capitalizeWords(words: string) {
  return typeof words === 'string' ? words.replace(/(?:^|\s)\S/g, (w) => w.toUpperCase()) : '';
}

/**
 * Capitalize all characters of a string
 *
 * @param {string} words String containing the words
 */
export function uppercase(words: string): string {
  return typeof words === 'string' ? words.toUpperCase() : '';
}

/**
 * Ellipsize a string if it has more than n characters
 *
 * @param word String to ellipsize
 * @param maxLength Max length of the string
 * @param ellipsizeChars Characters to use in the ellipsis
 */
export function ellipsize(word: string, maxLength: number, ellipsizeChars = '...'): string {
  return word.length > maxLength ? word.substring(0, maxLength - 3) + ellipsizeChars : word;
}
