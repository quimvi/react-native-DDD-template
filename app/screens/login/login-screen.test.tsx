import * as React from 'react';

import { MockApplication, IMockApplication } from '~/application';
import { RequestSubject, Status } from '~/observables';
import {
  fireEvent,
  RenderAPI,
  waitFor,
  act,
  renderWithApplicationLayer,
} from '~/test/utils';
import { MockNavigator } from '~/test/utils/mock-navigator';

import { LoginScreen } from './login-screen';

jest.mock('jwt-decode', () => jest.fn());

describe('LoginScreen', function () {
  let instance: RenderAPI;
  let navigate: jest.Mock;
  let application: IMockApplication;

  beforeEach(function () {
    navigate = jest.fn();
    application = new MockApplication();
    instance = renderWithApplicationLayer(
      <MockNavigator navigate={navigate}>
        <LoginScreen />
      </MockNavigator>,
      application
    );
  });

  it('before the user press the login button', function () {
    expect(instance.toJSON()).toMatchSnapshot();
  });

  it('when the login in is in process', function () {
    act(() => {
      application.useCasesRequests.set('Login', new RequestSubject());
      application.useCasesRequests.get('Login').startRequest();
    });

    expect(instance.toJSON()).toMatchSnapshot();
  });

  it('when the login process finish with a success', async function () {
    fireEvent(instance.getByText(/send/i), 'press');
    await waitFor(() =>
      expect(
        application.useCasesRequests.get('Login')?.hasCompleted
      ).toBeTruthy()
    );

    expect(instance.toJSON()).toMatchSnapshot();
  });

  it('redirects to route after the login process finish', async function () {
    fireEvent(instance.getByText(/send/i), 'press');
    await waitFor(() =>
      expect(application.useCasesRequests.get('Login')?.requestStatus).toBe(
        Status.DONE
      )
    );

    expect(navigate).toHaveBeenCalledWith('Main');
  });
});
