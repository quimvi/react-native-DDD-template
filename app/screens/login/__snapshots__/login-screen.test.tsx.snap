// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen before the user press the login button 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Send
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`LoginScreen when the login in is in process 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Send
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`LoginScreen when the login process finish with a success 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Send
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
